File: server/Simulation/Systems/LifetimeSystem.cs
using server.ECS;
using server.Simulation.Components;

namespace server.Simulation.Systems;

/// <summary>
/// Manages entities with limited lifetimes (e.g., bullets, drops, temporary effects).
/// Tags entities for death when their lifetime expires.
/// </summary>
public sealed class LifetimeSystem : NttSystem<LifeTimeComponent>
{
    public LifetimeSystem() : base("Lifetime System", threads: 1) { }

    public override void Update(in NTT ntt, ref LifeTimeComponent c1)
    {
        c1.LifeTimeSeconds -= DeltaTime;

        if (c1.LifeTimeSeconds > 0)
            return;

        var dtc = new DeathTagComponent(default);
        ntt.Set(ref dtc);
    }
}

File: pixiejsClient/src/app/ecs/systems/LifetimeSystem.ts
import { System1 } from "../core/System";
import { NTT } from "../core/NTT";
import { LifeTimeComponent } from "../components/LifeTimeComponent";
import { DeathTagComponent } from "../components/DeathTagComponent";

export class LifetimeSystem extends System1<LifeTimeComponent> {
  constructor() {
    super(LifeTimeComponent);
  }

  protected updateEntity(ntt: NTT, ltc: LifeTimeComponent, deltaTime: number): void {
    ltc.lifetimeSeconds -= deltaTime;

    if (ltc.lifetimeSeconds > 0)
      return

    ntt.set(new DeathTagComponent(ntt));
  }
}
